### ISSUE TITLE
JsonRpc 2.0 array requests

### ISSUE THREAD

**Issue by keldu** [capnproto/capnproto#914]
The Json Rpc class doesn't seem to be able to make jsonrpc2.0 array requests as far as I can see it. At least there are array requests according to [specification](https://www.jsonrpc.org/specification).  

I have some code hacked together which is based on capnproto and rewrote parts of the jsonrpc class and made it usable in the `kj::HttpService` class at "github.com/keldu/genralapi". But since I changed to much, have added unnecessary/unfinished stuff and don't have a client implemented I don't want to merge my code into capnproto. Also it's not written to my standard, but I can make jsonrpc2.0 array calls. I can write a proper merge though.  

I just wanted to open an issue before doing that.  

My hacky implementation generates orphans if an array request was sent and adopts them into an appropriate array of orphans and serializes that into json, mainly because I don't know the response size since notifications don't generate array elements.

**Comment by keldu** [comment]
I find it kind of unsettling that the JsonRpc class is not easily combinded with the `kj::HttpService` class.  
But I can see that this change is too big of an impact for the people who already use this functionality.

**Comment by kentonv** [comment]
HI @keldu,

The JSON-RPC implementation is very basic right now as I wrote it only for a very specific use case that I had (implementing a language server for VS Code). That use case used a raw socket transport, not HTTP, so I didn't implement HTTP integration. I'd definitely love to see the code extended to support other use cases.

How exactly does JSON-RPC work over HTTP, given that HTTP is strictly request/response while JSON-RPC is asynchronous bidirectional? Do you use WebSocket, or do you just say that in this case the interactions are limited to request/response?

**Comment by keldu** [comment]
Hey @kentonv ,

now I see why you did implement it this way.  
the jsonrpc1.0 has a peer-peer model, but jsonrpc2.0 has a client-server model. At least that's what the [website](https://www.jsonrpc.org/specification) says. The jsonrpc2.0 server may still handle jsonrpc1.0 messages with the obvious client-server restriction according to the Compability section.   

So that's why I used a simple `kj::Promise<kj::String>(kj::String)` function for the request interpretation. I simply created a child of `kj::HttpService` and gave the Interpreter class to that child, so that the jsonrpc2.0 stays transport agnostic.

**Comment by kentonv** [comment]
Hmm, this seems confusing. The VS Code Language Server Protocol is very much JSON-RPC 2, but also uses bidirectional communications (the "server" often sends notifications to the client asynchronously). The JSON-RPC 2 spec doesn't seem say whether this is allowed or not -- they define "client" and "server" in terms of a single request, but nowhere in the spec does it seem to say whether you can both initiate and receive requests over the same connection (i.e. act as client for some requests and server for others). But LSP does in fact do this and hence my implementation supports it.

**Comment by keldu** [comment]
Oh no....  
I understand. Hmm.  
It does seem strange and I don't see an explicit restriction. It would make http connection impossible in a strict client-server modell, since that client can't initiate a connection, so I would guess that it is implicitly forbidden.  
But I'm not very sure myself anymore. I would like to have a proper RFC.
Restricting it to a connection would be easier to implement. You still can multiplex over one connection with some meta protocol.  

But ok, I still can implement batch/array calls, since that implementation isn't dependent on the peer-peer/client-server nature of JSON-RPC 2.0.

**Comment by kentonv** [comment]
I don't think you can even do it based on connections really. An HTTP connection might go to a proxy, which might split each request into a separate upstream connection, losing any connection state. Fundamentally, each HTTP request/response pair is its own unit independent of all others.

I guess the right way to think about a basic HTTP transport, then, is like you're creating a new `JsonRpc::Transport` for each request (or batch of requests), which can return exactly one response (or batch).

I haven't looked very closely myself but it seems likely that it would be difficult to really integrate this with the existing `JsonRpc` implementation, which is really meant for streaming transports. It might indeed make sense to think of JSON-RPC-over-HTTP as a completely separate thing with its own implementation.

**Comment by keldu** [comment]
JSON-RPC-2.0 makes it clear that it wants to be transport independent contrary to JSON-RPC-1.0, which allows this HTTP distinction.
At least in HTTP I did it as you just said by creating a single connection per single/batch request.

But I agree that this would be hard to integrate in your current implementation.
